File: yumin/24.11.14/1027.py
# 시간 제한이 2초, 진하게 느껴지는 브루트포스의 냄새..
# 건물과 건물 사이의 기울기에 관한 문제
# 두 점 사이를 지나는 직선의 방정식 공식을 썼다.

n = int(input())
buildings = list(map(int, input().split()))
cnt = 0

for i in range(n):
    temp = n - 1 # 기준이 되는 건물을 제외한 수에서 하나씩 차감할 예정

    for j in range(i + 1, n): # 기준이 되는 건물의 오른쪽 부분 (기울기 증가를 체크)
        for k in range(i + 1, j):
            if buildings[k] - buildings[i] >= ((buildings[j] - buildings[i]) / (j - i)) * (k - i): # 건물과 건물 사이에, 더 큰 건물이 있다면
                temp -= 1 # 다른 건물이 안보이니까 차감
                break
        
    for j in range(0, i): # 기준이 되는 건물 왼쪽 부분 (기울기 감소를 체크)
        for k in range(j + 1, i):
            if buildings[k] - buildings[j] >= ((buildings[i] - buildings[j]) / (i - j)) * (k - j): # 건물과 건물 사이에, 더 큰 건물이 있다면
                temp -= 1 # 다른 건물이 안보이니까 차감
                break
    
    cnt = max(temp, cnt) # 제일 많이 보이는 건물의 수 골라내기

print(cnt)

File: dongha/241114/1027.cpp
#include <algorithm>
#include <cmath>
#include <iostream>
#include <map>
#include <queue>
#include <set>
#include <sstream>
#include <vector>
#define PATH "/Users/leedongha/Downloads/PS/input.txt"
#define L_PATH "input.txt"
#define fastio cin.tie(0)->sync_with_stdio(0);
#define rep(n) for (int i = 0; i < n; ++i)
#define si(x) int(x.size())
#define all(x) (x).begin(), (x).end()
#define pb(...) push_back(__VA_ARGS__)
#define X first
#define Y second
#define ROOT 1
#define INF 0x3f3f3f3f
using namespace std;
using ll = long long;
using TP = tuple<int, int, int>;
using P = pair<int, int>;

inline ll cross(ll x1, ll y1, ll x2, ll y2) { return x1 * y2 - x2 * y1; }

ll n, arr[51];
int main() {
  fastio;
  cin >> n;
  for (int i = 0; i < n; ++i) cin >> arr[i];
  int ans = 0;

  for (ll ax, ay, bx, by, i = 0; i < n; ++i) {
    int cnt = 0, cntt = 0;
    for (ll j = i - 1; j >= 0; --j) {
      bool f = 1;
      ax = i;
      ay = arr[i];
      bx = j;
      by = arr[j];

      for (ll cx, cy, dx, dy, k = i - 1; k > j; --k) {
        cx = dx = k;
        cy = 0;
        dy = arr[k];
        ll d1 = cross(bx - ax, by - ay, cx - ax, cy - ay), d2 = cross(bx - ax, by - ay, dx - ax, dy - ay);
        if (!(d1 > 0 && d2 > 0 || d1 < 0 && d2 < 0 || d1 == 0 && d2 == 0)) {
          f = 0;
          break;
        }
      }

      if (!f) continue;
      cnt++;
    }

    cntt += cnt;
    cnt = 0;

    for (ll j = i + 1; j < n; ++j) {
      bool f = 1;
      ax = i;
      ay = arr[i];
      bx = j;
      by = arr[j];

      for (ll cx, cy, dx, dy, k = i + 1; k < j; ++k) {
        cx = dx = k;
        cy = 0;
        dy = arr[k];
        ll d1 = cross(bx - ax, by - ay, cx - ax, cy - ay), d2 = cross(bx - ax, by - ay, dx - ax, dy - ay);
        if (!(d1 > 0 && d2 > 0 || d1 < 0 && d2 < 0 || d1 == 0 && d2 == 0)) {
          f = 0;
          break;
        }
      }

      if (!f) continue;
      cnt++;
    }

    cntt += cnt;
    ans = max(ans, cntt);
  }

  cout << ans << "\n";
}